Migration to webpack

b4dnewz opened the pull request:
### Project build

A little recap, grunt is old and slower compared to webpack so I proposed to migrate to it.
This is the build setup using webpack, I've also added the browser field to package.json file pointing to the unminified build for the web projects, it's not ready yet to fully replace grunt we must move the documentation generator also... I've tried also to replicate the grunt-documentation plugin using the documentation npm module but the output is quite different and I don't understand why since the options are the same, probably it's because I'm using a newer version.

---

### Project documentation

TL;DR [DEMO](https://core.codekraft.it/public/docs/php-parser/)

This is [the file](https://github.com/b4dnewz/php-parser/blob/migration/documentation/build-docs.js) I've made for programmatical use of documentationjs, but the __output is somehow different__ :( probably depends of the newer version, but I don't know exactly.

I've created [another branch](https://github.com/b4dnewz/php-parser/tree/migration/documentation) to try out __jsdoc__ with this project and managed to make it work after fixing a couple of tags errors, there are still some checks to do, regards examples and other parts of the documentation to ensure they match the jsdoc standard... but [this](https://core.codekraft.it/public/docs/php-parser/) is a preview of the result.

This is the default theme, if you like this solution we can use the theme you like more.

I think in terms of readability and usability is better than the wiki pages, I admit I'm used to read stuff in github, I spend most of my time on it, but since jsdoc (esdoc, tsdoc and friends) are widely used I feel familiar to move around the documentation generated by that tools.

---

The idea is to migrate the documentation, merge it in the webpack branch and than merge back into v3.1.
What do you think about it?

b4dnewz commented:
I've restored the old grunt/grunt-documentation deps and tried to generate the documentation and the output is different, also I've tried to checkout the master branch as generate documentation and the output is still different so I guess the documentation is __just old__ and there are no changes in the plugin, here a sample of a diff:

![image](https://user-images.githubusercontent.com/29862596/42501422-03e77a2a-8434-11e8-90e6-a267e704acc8.png)

I guess using [this file](https://github.com/b4dnewz/php-parser/blob/migration/documentation/build-docs.js) is cool and I can remove grunt and use [documentationjs node api](https://github.com/documentationjs/documentation/blob/master/docs/NODE_API.md#formatsmarkdown) to replace the grunt task.

ichiriac commented:
Hi @b4dnewz, I like the JSDoc output but when I have to read that kind of documentation the hard part is that everything is mixed and information is condensed. 

I want to provide a simple quick start and don't bother the reader with too technical information about internal functions from the API.

Maybe by using another template could help, like this one https://github.com/braintree/jsdoc-template/ or this one : https://github.com/docstrap/docstrap#readme

Do we can add cookbook pages on JSDoc in order to also provide some mini tutorials ?

If the html output is clean and readable, we may give up the markdown version,  no more need to maintain it.

ichiriac commented on .npmignore:
```diff
@@ -1,9 +1,10 @@
 /.istanbul.yml
-/.travis.yml
\ No newline at end of file
+/.travis.yml
```
:+1:

ichiriac commented on example/index.html:
```diff
@@ -0,0 +1,114 @@
+              </a>
+          </div>
+          <div class="md-toolbar-section-end">
+            <md-checkbox v-model="live" true-value="true" false-value="false">Live Mode</md-checkbox>
```
could you provide extra options ;
- Positions
- Comments
- Extract the source
- Extract tokens
- Failover on errors

ichiriac commented on example/index.html:
```diff
@@ -0,0 +1,114 @@
+            </md-field>
+          </div>
+          <div class="md-layout-item md-size-40 output-container">
+            <pre v-text="output"></pre>
```
could you use a JSON formatter like for example https://myst729.github.io/vue-json-tree/ ? 

For huge outputs it should be simpler to navigate

ichiriac replied:
BTW, I really like to demo, I've tried it online here : https://rawgit.com/b4dnewz/php-parser/migration/webpack/example/index.html

ichiriac commented on package.json:
```diff
@@ -26,17 +31,31 @@
   "author": "Ioan CHIRIAC",
+  "contributors": [
+    {
+      "name": "Filippo Conti",
```
:+1: :heart:

b4dnewz commented:
hi @ichiriac good morning, I was about to write you, just few minutes because I found in the documentation the tutorials give a 404 caused from a bad href reference

ichiriac commented:
hi @b4dnewz, good morning :)

I really like the new demo, you have done a great job here !

b4dnewz commented:
I've this strange problem now that is making me crazy I don't now why the tutorial files generated by jsdoc are capitalized but when I go to commit them they are lower cased O_o which results in a 404 on documentation href

![image](https://user-images.githubusercontent.com/29862596/43364191-f972aba0-9315-11e8-8406-bd862d78ab5b.png) 
![image](https://user-images.githubusercontent.com/29862596/43364200-19010ae8-9316-11e8-868a-5ba549ed0ac0.png)

mysteries of life...

b4dnewz commented:
Ok.. I'm not exactly sure how to fix codeclimate I though adding docs folder to eslintignore would fix it but I was wrong.

Talking about the PR this is was I made:

I've updated the [demo](https://rawgit.com/b4dnewz/php-parser/migration/webpack/example/index.html) page including all the parser settings, changed the jsdoc documentation [template](https://rawgit.com/b4dnewz/php-parser/migration/webpack/docs/index.html) but as you can see it still needs some customization.

I've added some tutorial pages and references [here](https://rawgit.com/b4dnewz/php-parser/migration/webpack/docs/tutorial-Parser_.html) an example.

It still need to be completed updating the comments pointing the various parts of the code to their relative example/tutorial if any, I took these from the old docs folder.

This branch is behind the master few commits I try to align them and then we can merge if nothing fails.

ichiriac commented on .codeclimate.yml:
```diff
@@ -0,0 +1,3 @@
+exclude_patterns:
```
:+1:

ichiriac replied:
all add `dist/` folder

b4dnewz commented:
Oh jesuss... now I add also test folder

b4dnewz commented:
Ok I guess if we ignore the codeclimate warning it's ready.

ichiriac commented:
It's strange because index.d.ts is not modified by your PR so it should not be considered as a new issue, and also it's a false positive as it's not a duplicate code :

![image](https://user-images.githubusercontent.com/173203/43364559-cd1bdf0c-931c-11e8-9572-acaa90177a4a.png)

We are ready to go :100:

ichiriac commented:
Hi @b4dnewz,

I've merged your PR, and as the demo is really cool so I've published it on glayzzle. You have access over https://github.com/glayzzle/glayzzle.github.io - there you can publish changes as you want

I think it would be better that the `example` and `docs` folders are published on the website repository - so I've removed/ignored them from the repository. 

In order to release the docs we can run the generatio from this repository and next we copy/paste the docs folder on the website repository, unless you know a better way to achieve this.

b4dnewz commented:
hi @ichiriac, usually I create a custom `package.json` script with the following command:
```
git subtree push --prefix demo origin gh-pages`
```
For example _package.json_:
```json
"scripts": {
  "build": "webpack --config webpack.config.js",
  "publish-docs": "git subtree push --prefix docs origin gh-pages"
},
```
Once we publish a new version or update the docs we run npm run **publish-docs** and we have the documentation hosted on: https://glayzzle.github.com/php-parser

Do you like that?

b4dnewz commented:
Than for the official website the separate repository it's ok and the Documentation link can point to the php-parser gh-pages

ichiriac commented:
Yes, it's great, I like it ! This way we'll be able to publish on the various packages :+1:

b4dnewz commented:
yup, later this evening I'll add the command and publish the first gh-pages with the documentation, ok?
do you want to check some of the markdown tutorials first?

ichiriac commented:
I have to rewrite them, but for a first shot it's OK. When your system will be released it will be easy to update them. You can go ahead and move them to gh-pages

Really thanks for your help, without you I'll never get the time to do that sort of user-frendly improvements

b4dnewz commented:
no worries, when I have time is always a pleasure to help

b4dnewz commented:
@ichiriac i'm proceeding, in order to use git subtree we __must__ track`docs` folder into git and commit it, i've updated the package.json with the **files** property to include only `lib `and `dist` folder so the npm package remains slim, while the github repository is bigger because has all files in it and the gh-pages only the last committed `docs` folder while subtreeing.

Also, I'm going to setup [Algolia DocSearch](https://community.algolia.com/docsearch/) to enable documentation search as suggested from [jsdoc-template](https://github.com/braintree/jsdoc-template#example-jsdoc-config), which right now is missing and is quite difficult to navigate the classes, when I've all set up I set your as the owner of that so you can manage and maybe in the future we can use for more documentations, is free (it says)

For now I just subtree push the docs as it is until I've configured everything.

ichiriac commented:
yes of course, no problem

b4dnewz commented:
Ok I made it, but you probably have to set/check the gh-pages branch in the repository settings if the 404 doesn't go away in few minutes https://glayzzle.com/php-parser

ichiriac commented:
Hi @b4dnewz,

The search engine is really helpfull and I like the result display. I hope the opensource plan will be enough but anyway, it's a great feature.

Do you think we could integrate the playground demo as a custom page of the documentation ?

I've forked the template here https://github.com/glayzzle/jsdoc-template, so we will be able to customise or fix css if needed.

b4dnewz commented:
> Do you think we could integrate the playground demo as a custom page of the documentation ?

Good idea, I will try
